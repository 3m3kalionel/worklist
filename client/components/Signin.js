import React from 'react';
import { Mutation } from 'react-apollo';
import Link from 'next/link';

import AuthFormStyles from '../styles/AuthFormStyles';
import { WithForm } from './HOC';

import SIGN_IN_MUTATION from '../graphql/mutations/Signin';

const Signin = props => {
  const { onChange, onSubmit, onError, values } = props;
  const { email, password } = values;

  return (
    <Mutation
      mutation={SIGN_IN_MUTATION}
      variables={{ email, password }}
    >
      {(signin, { error, loading }) => (
        <form
          onSubmit={onSubmit(signin, 'signin')}
        >
          <AuthFormStyles>
            <div>
              <div id="authentication-form">
                <div id="form-body">
                  <h1 id="form-header">Welcome Back</h1>
                  <label htmlFor="email">Email</label>
                  <input
                    name="email"
                    type="email"
                    className="auth-form-input"
                    placeholder="Email"
                    value={email}
                    autoComplete="email"
                    onChange={onChange}
                    required
                  />
                  <label htmlFor="password">Password</label>
                  <input
                    name="password"
                    type="password"
                    className="auth-form-input"
                    placeholder="Password"
                    value={password}
                    autoComplete="current-password"
                    onChange={onChange}
                    required
                  />
                  <div id="action-button-container">
                    <button type="submit" id="sign-up-button">Login</button>
                    <span
                      id="forgot-password"
                    >
                    <Link href="/requestreset">
                      <a>Forgot Password?</a>
                    </Link>
                  </span>
                  </div>
                </div>
              </div>
            </div>
          </AuthFormStyles>
        </form>
      )}
    </Mutation>
  )
}

export default WithForm(Signin);
