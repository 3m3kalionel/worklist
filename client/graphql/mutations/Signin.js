import gql from 'graphql-tag';

const SIGN_IN_MUTATION = gql`
  mutation SIGN_IN_MUTATION($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      id
      email
      username
      authoredTodoLists {
        id
      }
      todoListsContributedTo {
        id
      }
    }
  }
`;

export default SIGN_IN_MUTATION
