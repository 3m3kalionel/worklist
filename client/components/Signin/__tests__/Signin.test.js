import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ApolloProvider } from 'react-apollo';
import '@testing-library/react/cleanup-after-each';
import { createMockClient } from 'mock-apollo-client';

import Signin from '../Signin';

import SIGN_IN_MUTATION from '../../../graphql/mutations/Signin';

jest.mock("../../HOC.js", () => Component => {
  const mockHOC = props => (
    <div><Component {...props} /></div>
  );

   return mockHOC;
});

const setup = (values, onSubmit) => {
  const onSubmitHandler = onSubmit || jest.fn();
  const mockClient = createMockClient();

  mockClient.setRequestHandler(
    SIGN_IN_MUTATION,
    () => Promise.resolve({
      data: {
        signin: {
          id: "cjyu8h6467e380b53tn7ezb58",
          username: "emeka",
          email: "emeka.l.okoro@gmail.com",
          profilePicUrl: null,
          authoredTodoLists: null,
          todoListsContributedTo: null
        }
      }
    })
  );

  const {
    container,
    getByText,
    getByTestId
  } = render(
    <ApolloProvider client={mockClient}>
      <Signin
        values={values}
        onSubmit={onSubmitHandler}
        onChange={jest.fn()}
      />
    </ApolloProvider>
  );

  return { container, getByText, getByTestId }
}

describe('<Signin />', () => {
  test('renders an empty form', () => {
    const { container } = setup({
      email: "",
      password: ""
    });

    expect(container).toMatchSnapshot();
  });

  test('renders text in the form input', () => {
    const { container } = setup({
      usernname: "username",
      email: "email@gmail.com",
    });

    expect(container).toMatchSnapshot();
  });

  test('calls the submit handler when button is clicked', () => {
    const onSubmitHandler = jest.fn();
    const {
      container,
      getByTestId
    } = setup({
      usernname: "username",
      email: "email@gmail.com",
      password: "password",
      phoneNumber: "phoneNumber"
      },
      onSubmitHandler
    );
    const form = getByTestId('form');
    fireEvent.submit(form);
  
    expect(onSubmitHandler).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
})
