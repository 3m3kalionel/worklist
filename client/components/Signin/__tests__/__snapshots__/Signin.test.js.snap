// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signin /> calls the submit handler when button is clicked 1`] = `
<div>
  <div>
    <form
      data-testid="form"
    >
      <div
        class="AuthFormStyles-js87jd-0 ieHHDD"
      >
        <div>
          <div
            id="authentication-form"
          >
            <div
              id="form-body"
            >
              <h1
                id="form-header"
              >
                Welcome Back
              </h1>
              <label
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="auth-form-input"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value="email@gmail.com"
              />
              <label
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="auth-form-input"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value="password"
              />
              <div
                id="action-button-container"
              >
                <button
                  id="sign-up-button"
                  type="submit"
                >
                  Login
                </button>
                <span
                  id="forgot-password"
                >
                  <a
                    href="/forgotpassword"
                  >
                    Forgot Password?
                  </a>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`<Signin /> renders an empty form 1`] = `
<div>
  <div>
    <form
      data-testid="form"
    >
      <div
        class="AuthFormStyles-js87jd-0 ieHHDD"
      >
        <div>
          <div
            id="authentication-form"
          >
            <div
              id="form-body"
            >
              <h1
                id="form-header"
              >
                Welcome Back
              </h1>
              <label
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="auth-form-input"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value=""
              />
              <label
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="auth-form-input"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
              <div
                id="action-button-container"
              >
                <button
                  id="sign-up-button"
                  type="submit"
                >
                  Login
                </button>
                <span
                  id="forgot-password"
                >
                  <a
                    href="/forgotpassword"
                  >
                    Forgot Password?
                  </a>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`<Signin /> renders text in the form input 1`] = `
<div>
  <div>
    <form
      data-testid="form"
    >
      <div
        class="AuthFormStyles-js87jd-0 ieHHDD"
      >
        <div>
          <div
            id="authentication-form"
          >
            <div
              id="form-body"
            >
              <h1
                id="form-header"
              >
                Welcome Back
              </h1>
              <label
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="auth-form-input"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value="email@gmail.com"
              />
              <label
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="auth-form-input"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
              <div
                id="action-button-container"
              >
                <button
                  id="sign-up-button"
                  type="submit"
                >
                  Login
                </button>
                <span
                  id="forgot-password"
                >
                  <a
                    href="/forgotpassword"
                  >
                    Forgot Password?
                  </a>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
